Add onReceiveCIS2 Hook Tutorial (#1287)
Concordium/concordium.github.io: source/mainnet/tutorials/onReceivingCIS2/on-receivingCIS2.rst

When one contract sends CIS-2 tokens to another contract, the receiving contract needs a way to handle these incoming tokens properly. This is where the ``onReceivingCIS2`` hook comes in.

This mechanism enables contracts to process incoming tokens correctly. Think of the ``onReceivingCIS2`` function like a building's mailroom. Just as a mailroom has specific procedures for receiving, sorting, and distributing packages, this function defines how your contract handles incoming tokens. \
When another contract sends tokens to your contract, the ``onReceivingCIS2`` hook automatically triggers, allowing your contract to process these tokens according to your defined rules.

This hook is essential because:

- It prevents tokens from becoming permanently locked
- It lets you define specific handling logic for incoming tokens
- It enables secure contract-to-contract token transfers

Creating a token forwarder contract
====================================

Follow these steps to create a contract that automatically forwards any received tokens to its owner. This pattern helps protect against tokens becoming stuck in your contract.

Initialize the Project
----------------------

> Create a new smart contract project using the **init** command. The project will be called **token-receiver**. Then add the **concordium-cis2** library.

Reviewer: If the reader should use the `cargo concordium init` command then the next step would be to explain which template should be selected among the options. E.g.
```
After selecting template `xyz`, the command line will ask for a project name. Call it `token-receiver` to follow along with this tutorial. 
```

Author: Sentence modified accordingly.

        fn from(_: CallContractError<T>) -> Self {
            Self::InvokeContractError
        }
    }

    type ContractError = Cis2Error<CustomContractError>;

A simple message string will be stored in the state.

3. Modify the initialization function:

.. code-block:: rust

    #[init(contract = "token_forwarder")]
    fn init(_ctx: &InitContext, _state_builder: &mut StateBuilder) -> InitResult<State> {
        Ok(State {
            message: "Token Forwarder initialized!".to_string(),
        })
    }

This method will initialize the contract and assign a value to the **message string**.

4. Modify the token receiving function. This will become the main method of the contract, the ``onReceivingCIS2`` function that handles incoming tokens:

.. code-block:: rust

    #[receive(
            contract = "token_forwarder",
            name = "onReceivingCIS2",
            parameter = "OnReceivingCis2Params<ContractTokenId, ContractTokenAmount>",
            error = "ContractError"
        )]
        fn contract_receive_cis2(ctx: &ReceiveContext, host: &Host<State>) -> Result<(), ContractError> {
            // Get information about received tokens

>             let params: OnReceivingCis2Params<ContractTokenId, ContractTokenAmount> = ctx.parameter_cursor().get()?;

Reviewer: Sind you introduce this `message` string in the state. I suggest to either remove it or if it should help the reader to understand how the contract works then the `onReceivingCIS2` function should probably also update the string to something like "Forworded tokenId xyz from contract xyz to owner".

Author: I've added the message:
 "Forwarded tokenId {:?} from contract with index {} to owner".
Please check if that's done properly and resolve this comment if so.

                _ => return Ok(()), // Non-contract senders are ignored
            };

            // Create transfer to forward tokens to contract owner
            let transfer = Transfer {
                token_id: params.token_id,
                amount: params.amount,
                from: Address::Contract(ctx.self_address()),
                to: Receiver::from_account(ctx.owner()),
                data: AdditionalData::empty(),
            };

            // Execute the transfer
            host.invoke_contract_read_only(
                &token_contract,
                &TransferParams::from(vec![transfer]),
                "transfer",
                Amount::zero(),
            )?;

            Ok(())
        }

5. Add helper functions to manage the state:

.. code-block:: rust

    #[receive(
        contract = "token_forwarder",
        name = "set_message",
        parameter = "String",
        error = "ContractError",
        mutable
    )]
    fn set_message(ctx: &ReceiveContext, host: &mut Host<State>) -> Result<(), ContractError> {
        let new_message: String = ctx.parameter_cursor().get()?;
        host.state_mut().message = new_message;
        Ok(())

>     }

Reviewer: To keep contract code simple, I feel we don't need this function around manipulating the message strings.

    #[receive(
        contract = "token_forwarder",
        name = "view",
        return_value = "String",
        error = "ContractError"
    )]
    fn view(_ctx: &ReceiveContext, host: &Host<State>) -> Result<String, ContractError> {
        Ok(host.state().message.clone())
    }

These helper methods provide the ability to modify and view the message stored in the state.

Define the Contract Schema
--------------------------

For our contract to be properly defined on the blockchain, we recommend creating a schema. Create a file named ``schema.json`` in your project root, it will specify the parameter and error types for the contract methods.:

> 

Reviewer: Schema files should not be typed by hand. They can be generated by `cargo concordium`:
```
cargo concordium build --schema-json-out ./
```
or 
```
cargo concordium build --schema-template-out -
```

Since you don't use the `schema.json` file anywhere in the tutorial, the section `define the contract schema` can be removed.

Author: I've removed the `define the contract schema` section.

            }
        },
        "set_message": {
            "parameter": "String",
            "error": {
                "type": "enum",
                "variants": {
                    "ParseParams": {
                        "type": "unit"
                    },
                    "InvokeContractError": {
                        "type": "unit"
                    }
                }
            }
        },
        "view": {
            "returnValue": "String",
            "error": {
                "type": "enum",
                "variants": {
                    "ParseParams": {
                        "type": "unit"
                    },
                    "InvokeContractError": {
                        "type": "unit"
                    }
                }
            }
        }
    }

Build and Deploy
----------------

Now let's build our contract:

.. code-block:: console

    $ cargo concordium build --out tokenForwarder.module.wasm.v1

After building successfully, a module file will be created. Next, let's deploy the contract:

.. code-block:: console

    $ concordium-client module deploy tokenForwarder.module.wasm.v1 \
    --sender <your-account-address> \

>     --grpc-port 20000 --grpc-ip node.testnet.concordium.com \

Reviewer: Please use the `grpc.testnet.concordium.com` endpoint which are several nodes behind a loadbalancer. The `node.testnet.concordium.com` node was just setup for a hackathon and might be removed.

Note: the `grpc.testnet.concordium.com` endpoint uses TLS and needs `--secure` flag added to your command

Author: I've switched to `grpc.testnet.concordium.com` and added the `--secure` flag

After successful deployment, you'll receive a module reference in the following format:

.. code-block:: json
    :force:

    d121f262f3d34b9737faa5ded2135cf0b994c9c32fe90d7f11fae7cd31441e86

Save this reference - you'll need it for contract initialization and future reference. To initialize, run the following command:

.. code-block:: console

    $ concordium-client --grpc-port 20000 --grpc-ip node.testnet.concordium.com \
    contract init <saved_reference> \
    --sender <your-account-address> \
    --contract token_forwarder \

>     --energy <max-energy-allowed> \

Reviewer: please align the command with the previous command (e.g. have  `grpc-port` and `grpc-ip` at the beginning or the end always)

Author: All the commands are aligned now.

If successful, you will receive a message with the contract's index and subindex, in the following format:

.. code-block:: json
    :force:

    Contract successfully initialized with address: {"index":10710,"subindex":0}

Testing the functionality
-------------------------

We will test our token forwarder on the testnet. We'll need some **CIS-2 tokens** (like wCCD) and we'll send them to the **token forwarder**, using the `wCCD <https://github.com/Concordium/concordium-rust-smart-contracts/blob/main/examples/cis2-wccd/src/lib.rs>`_ contract as an entrypoint, such that the transfer will be sent from contract to contract.

> Before continuing, you need to use the `wCCD dApp <https://wccd.testnet.concordium.com/>`_ to wrap your **CCD**.

Reviewer: ```suggestion
You can acquire some wCCD tokens on the `wCCD dApp <https://wccd.testnet.concordium.com/>`_ by wrapping some of your **CCD**.
```

Author: Suggestion implemented in recent commits.

                                ENTRYPOINT_NAME
                            ]
                        }
                    ]
                },
                "token_id": TOKEN_ID
            }
        ]

    .. note::

        You can execute several transfers in the above array.

    If you insert everything correctly, the JSON array should look similar to
    the below JSON array that will transfer 1000 wCCD from an account address(your account in this case) to a contract.

    Replace **<token-forwarder-index>** with the index of the received in the previous section.

    .. code-block:: json
        :force:

        [
            {
            "amount": "1000",
            "data": "",
            "from": {
                "Account": [
                    "2xUeDsSS8fRfdXuPeDiFTsrJyxsAQhCw2xw9GYARnJpCM3iaeL"
                ]
            },
            "to": {
                "Contract": [
                    {
                        "index": <token-forwarder-index>,
                        "subindex": 0
                    },
                    "onReceivingCIS2"
                ]
            },
            "token_id": ""
            }
        ]

> The index of the **wCCD contract** is **2059**.

Reviewer: ```suggestion
The index of the **wCCD contract** is **2059** on testnet.
```

Author: Suggestion implemented in recent commits.
